### ISSUE TITLE
Accessibility Problem

### ISSUE THREAD

**Issue by kwfinken** [openseadragon/openseadragon#1478]
When using openseadragon with a screen reader and navigating by keyboard, the voice prompts are all "clickable", making it virtually useless.

Before someone says, we don't need to do it because people using screen readers are blind and don't need images, take note:  1) a large group of users of screen readers are not blind but low vision or functionally dyslexic.  2) The low vision group often zooms into images so that they can see them as a sighted person would without zooming in. 3) Another common user group for keyboards/screen readers are those with motor impairments who may be able to see perfectly fine.

**Comment by iangilman** [comment]
This would be great to fix! Thank you for drawing attention to it. 

Is this something you might be able to tackle? If not, can you provide some pointers for whoever might take it on?

**Comment by btzr-io** [comment]
Tested with firefox accessibility inspector:

### Canvas accessibility 
  > Text Labels and Names
 > Error: Content with images must be labeled. [learn more](https://developer.mozilla.org/en-US/docs/Web/Accessibility/Understanding_WCAG/Text_labels_and_names?utm_source=devtools&utm_medium=a11y-panel-checks-text-label#Content_with_images_must_be_labeled)

- Additional steps:  [Hit regions and accessibility](https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API/Tutorial/Hit_regions_and_accessibility)

### Controls accessibility
> A toolbar is a container for grouping a set of controls, such as buttons, menubuttons, or checkboxes.
> When a set of controls is visually presented as a group, the toolbar role can be used to communicate the presence and purpose of the grouping to screen reader users. Grouping controls into toolbars can also be an effective way of reducing the number of tab stops in the keyboard interface. 

See: https://www.w3.org/TR/wai-aria-practices-1.1/#toolbar

**Comment by msalsbery** [comment]
Hey, looks like we found someone to tackle accessibility issues! I’ll be looking forward to the pull request... ;)

**Comment by iangilman** [comment]
Yes, welcome, @btzr-io! [Villain](https://github.com/btzr-io/Villain/) looks like a cool project... I look forward to seeing it in action!

...and of course if you want to contribute accessibility fixes to OSD, they would be most welcome :)

**Comment by btzr-io** [comment]
thanks I'll see if I can help :+1:

**Comment by btzr-io** [comment]
Why there is so many wrappers for the control buttons ?
``` html
<div>
  <div>
    <div>
      { controls... }.
    </div>
  <div>
</div>
```
Here is a minimal markup for accessibility implementation:

```HTML
<div role="toolbar" aria-label="'Controls" aria-orientation="horizontal">
  <div role="button"  tabindex="0" .../>
  <div role="button" tabindex="-1" .../>
  ...
</div>
```
Focus management needs to be handled with `js`

> Implement focus management so the keyboard tab sequence includes one stop for the toolbar and arrow keys move focus among the controls in the toolbar.

**Comment by btzr-io** [comment]
I guess this is where the toolbar is created:
https://github.com/openseadragon/openseadragon/blob/68e9efebe699c5c47f274b572a9409d267bafaa7/src/viewer.js#L316-L318

Is `this.element` the main container of the toolbar ?
https://github.com/openseadragon/openseadragon/blob/68e9efebe699c5c47f274b572a9409d267bafaa7/src/controldock.js#L59-L65

**Comment by btzr-io** [comment]
### Workaround

- Create a your own custom [controls](https://openseadragon.github.io/examples/ui-binding-custom-buttons/)

- Assign and update `aria-label` attribute of canvas when navigating images:

```js
this.viewer.canvas.setAttribute('aria-label', "description...")
```

> Note: I'll update this as I work this issues on my project.

**Comment by iangilman** [comment]
> Why there is so many wrappers for the control buttons ?

I don't know for sure... Probably something to do with the various internal modules. Also, keep in mind that OpenSeadragon started life more than a decade ago, when IE6 was the standard browser. Presumably we had to do some things to get the right layout that aren't necessarily necessary today.

> I guess this is where the toolbar is created:

That's only if there's an external toolbar, like so:

http://openseadragon.github.io/examples/ui-toolbar/

Otherwise the viewer itself is the toolbar (note that `OpenSeadragon.Viewer` extends `OpenSeadragon.ControlDock`).

> Is `this.element` the main container of the toolbar ?

Seems likely. 

> Also what is this ID for ? I don't see it on the element.

Yeah, that appears to be vestigial; probably leftover from something that doesn't exist anymore.

> I think most of this issues can be handled externally without having to modify the source code:

Thank you for documenting these! Of course if you (or someone else) would like to help get them directly in to OSD (where applicable), that would be lovely too.
